### PR TITLE
Add support for different collection types

### PR DESCRIPTION
**Pull request by prokop7**
## Issue
`@CachableCollection` supports only Collection<?> arguments, it's not ideal for real use, often programmers use Set<?> or List<?>

## Proposed solution
Replace check `arg[0].class.equals(Collection.class)` with `Collection.class.isAssignableFrom(arg[0].class)`.

We have to decide what type of collection we have to create. The condition is that `arg[0].class` has to be assignable from this collection. To deal with it `CollectionCreator` interface is provided. Currently it consists two implementations: `DefaultCollectionCreator` which creates LinkedList, `SetCollectionCreator` that works with HashSet

### CODE REVIEW

File: collection-cacheable-for-spring-starter/src/test/java/de/qaware/tools/collectioncacheableforspring/CollectionCacheableTestRepository.java [hunk 1]
```diff
@@ -3,9 +3,7 @@
 import org.springframework.cache.annotation.CacheConfig;
 import org.springframework.cache.annotation.Cacheable;
 
-import java.util.Collection;
-import java.util.HashMap;
-import java.util.Map;
+import java.util.*;
```
Comment (neiser) [inline collection-cacheable-for-spring-starter/src/test/java/de/qaware/tools/collectioncacheableforspring/CollectionCacheableTestRepository.java:6]: Please don't use `*` imports, check your IntelliJ (or other IDE) settings

File: collection-cacheable-for-spring-starter/src/test/java/de/qaware/tools/collectioncacheableforspring/CollectionCacheableIntTest.java [hunk 2]
```diff
@@ -14,16 +14,10 @@
...
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
 
 import static de.qaware.tools.collectioncacheableforspring.CollectionCacheableTestRepository.CACHE_NAME;
-import static org.assertj.core.api.Assertions.assertThat;
-import static org.assertj.core.api.Assertions.entry;
+import static org.assertj.core.api.Assertions.*;
```
Comment (neiser) [inline collection-cacheable-for-spring-starter/src/test/java/de/qaware/tools/collectioncacheableforspring/CollectionCacheableIntTest.java:20]: No `*` imports.

File: collection-cacheable-for-spring-starter/src/main/java/de/qaware/tools/collectioncacheableforspring/creator/SetCollectionCreator.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+import java.util.Collection;
+import java.util.HashSet;
+
+@Component
+public class SetCollectionCreator
+        implements CollectionCreator {
+    @Override
+    public boolean canHandle(Class<?> cls) {
+        return cls.isAssignableFrom(HashSet.class);
```
Comment (neiser) [inline collection-cacheable-for-spring-starter/src/main/java/de/qaware/tools/collectioncacheableforspring/creator/SetCollectionCreator.java:33]: Shouldn't this be `equals`? What happens if somebody uses an argument with type which derives from HashSet? Can you add a test case for that?

**Comment by neiser** [comment]
@prokop7 Thanks for this PR. I'm willing to merge it as it looks reasonable. Can you look at my little review remarks?

**Comment by prokop7** [comment]
Also, I've fixed the misconfiguration of Cache Interceptor, it requires `factoryBean` to be able to get Caches

File: collection-cacheable-for-spring-starter/src/main/java/de/qaware/tools/collectioncacheableforspring/creator/SetCollectionCreator.java (hunk 3 above)
Comment (neiser) [reply collection-cacheable-for-spring-starter/src/main/java/de/qaware/tools/collectioncacheableforspring/creator/SetCollectionCreator.java:33]: Alright, thanks for clarification and the test case! I'll merge it and release.

**Review by neiser — APPROVED** [review]
Thanks for your work!

**Pull request merged**